cpp/src/rolling/detail/optimized_unbounded_window.cpp

```diff
@@ -0,0 +1,155 @@
+      // Note: cudf::reduce() does not support COUNT_ALL.
+      return cudf::make_count_aggregation<Base>(null_policy::INCLUDE);
+    } else if constexpr (std::is_same_v<Base, cudf::groupby_aggregation> and
+                         k == aggregation::Kind::COUNT_VALID) {
+      // Note: cudf::reduce() does not support COUNT_VALID.
+      return cudf::make_count_aggregation<Base>(null_policy::EXCLUDE);
+    } else if constexpr (k == aggregation::Kind::SUM) {
+      return cudf::make_sum_aggregation<Base>();
+    } else if constexpr (k == aggregation::Kind::MIN) {
+      return cudf::make_min_aggregation<Base>();
+    } else if constexpr (k == aggregation::Kind::MAX) {
+      return cudf::make_max_aggregation<Base>();
+    } else {
+      CUDF_FAIL("Unsupported aggregation kind for optimized unbounded windows.");
+    }
+  }
+};
+
+template <typename Base>
+std::unique_ptr<Base> convert_to(cudf::rolling_aggregation const& aggr)
+{
+  return cudf::detail::aggregation_dispatcher(aggr.kind, aggregation_converter<Base>{});
+}
+
+std::unique_ptr<column> aggregation_based_rolling_window(table_view const& group_keys,
+                                                         column_view const& input,
+                                                         rolling_aggregation const& aggr,
+                                                         rmm::cuda_stream_view stream,
+                                                         rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(group_keys.num_columns() > 0,
+               "Ungrouped rolling window not supported in aggregation path.");
+
+  auto agg_requests = std::vector<cudf::groupby::aggregation_request>{};
+  agg_requests.push_back(cudf::groupby::aggregation_request());
+  agg_requests.front().values = input;
+  agg_requests.front().aggregations.push_back(convert_to<cudf::groupby_aggregation>(aggr));
+
+  auto group_by = cudf::groupby::groupby{group_keys, cudf::null_policy::INCLUDE, cudf::sorted::YES};
+  auto aggregation_results           = group_by.aggregate(agg_requests, stream);
+  auto const& aggregation_result_col = aggregation_results.second.front().results.front();
+
+  using cudf::groupby::detail::sort::sort_groupby_helper;
+  auto helper = sort_groupby_helper{group_keys, cudf::null_policy::INCLUDE, cudf::sorted::YES, {}};
+  auto const& group_labels = helper.group_labels(stream);
+
+  auto result_columns = cudf::detail::gather(cudf::table_view{{*aggregation_result_col}},
+                                             group_labels,
+                                             cudf::out_of_bounds_policy::DONT_CHECK,
+                                             cudf::detail::negative_index_policy::NOT_ALLOWED,
+                                             stream,
+                                             mr)
+                          ->release();
+  return std::move(result_columns.front());
+}
+
+std::unique_ptr<column> reduction_based_rolling_window(column_view const& input,
```

Comment:
It would help to add a note that this code path is taking when there is no grouping so the rolling window is over the entire dataset. Right now neither this function nor the one above have any docs that I can see (including in the header since they're not declared there).

Reply:
Good point. Even if the functions are private, it would be best to document the intention.

Reply:
Documented.